File: exts/jphp-gui-ext/src/main/java/org/develnext/jphp/ext/javafx/classes/UXMediaPlayer.java
package org.develnext.jphp.ext.javafx.classes;

import javafx.scene.media.Media;
import javafx.scene.media.MediaPlayer;
import javafx.util.Duration;
import org.develnext.jphp.ext.javafx.JavaFXExtension;
import php.runtime.annotation.Reflection;
import php.runtime.annotation.Reflection.Getter;
import php.runtime.annotation.Reflection.Property;
import php.runtime.annotation.Reflection.Setter;
import php.runtime.annotation.Reflection.Signature;
import php.runtime.env.Environment;
import php.runtime.lang.BaseWrapper;
import php.runtime.reflection.ClassEntity;

@Reflection.Name(JavaFXExtension.NS + "UXMediaPlayer")
public class UXMediaPlayer extends BaseWrapper<MediaPlayer> {
    interface WrappedInterface {
        @Property double balance();
        @Property double rate();
        @Property double volume();
        @Property boolean mute();
        @Property MediaPlayer.Status status();

        @Property double currentRate();
        @Property int currentCount();
        @Property int cycleCount();

        void seek(Duration time);
        void play();
        void pause();
        void stop();
    }

    public UXMediaPlayer(Environment env, MediaPlayer wrappedObject) {
        super(env, wrappedObject);
    }

    public UXMediaPlayer(Environment env, ClassEntity clazz) {
        super(env, clazz);
    }

    @Signature
    public void __construct(Media media) {
        __wrappedObject = new MediaPlayer(media);
    }

    @Getter
    public Duration getCurrentTime() {
        return getWrappedObject().getCurrentTime();
    }

    @Setter
    public void setCurrentTime(Duration duration) {
        getWrappedObject().seek(duration);
    }

    @Getter
    public Media getMedia() {
        return getWrappedObject().getMedia();
    }

    @Getter
    public long getCurrentTimeAsPercent() {
        Duration duration = getWrappedObject().getMedia().getDuration();
        Duration currentTime = getWrappedObject().getCurrentTime();

        return Math.round(currentTime.toMillis() * 100.0 / duration.toMillis());
    }

    @Setter
    public void setCurrentTimeAsPercent(long percent) {
        if (percent > 100) {
            percent = 100;
        }

        Duration duration = getWrappedObject().getMedia().getDuration();

        getWrappedObject().seek(Duration.millis(Math.round(duration.toMillis() * percent / 100.0)));
    }
}

File: exts/jphp-gui-ext/src/main/resources/JPHP-INF/sdk/php/gui/UXMediaPlayer.php
<?php
namespace php\gui;


/**
 * Class UXMediaPlayer
 * @package php\gui
 * @packages gui, javafx
 */
class UXMediaPlayer
{
    /**
     * @var double form -1.0 to 1.0
     */
    public $balance = 0;

    /**
     * @var double from 0.0 to 8.0
     */
    public $rate = 1.0;

    /**
     * @var double
     */
    public $volume = 1.0;

    /**
     * @var bool
     */
    public $mute = false;

    /**
     * @var int -1 is
     */
    public $cycleCount = 1;

    /**
     * UNKNOWN, READY, PAUSED, PLAYING, STOPPED, STALLED, HALTED, DISPOSED
     * @readonly
     * @var string
     */
    public $status = 'UNKNOWN';

    /**
     * @readonly
     * @var double
     */
    public $currentRate;

    /**
     * @var int in millis
     */
    public $currentTime;

    /**
     * @var int in percent from 0 to 100
     */
    public $currentTimeAsPercent;

    /**
     * @readonly
     * @var int
     */
    public $currentCount;

    /**
     * @var UXMedia
     */
    public $media;

    /**
     * @param UXMedia $media
     */
    public function __construct(UXMedia $media)
    {
    }

    public function play()
    {
    }

    public function pause()
    {
    }

    public function stop()
    {
    }

    /**
     * @param $time in millis
     */
    public function seek($time)
    {
    }
}